// @ngxs/form-plugin/esm2015/src/actions.js
export class UpdateFormStatus {
    constructor(payload) {
        this.payload = payload;
    }
    static get type() {
        // NOTE: Not necessary to declare the type in this way in your code. See https://github.com/ngxs/store/pull/644#issuecomment-436003138
        return '[Forms] Update Form Status';
    }
}
export class UpdateFormValue {
    constructor(payload) {
        this.payload = payload;
    }
    static get type() {
        // NOTE: Not necessary to declare the type in this way in your code. See https://github.com/ngxs/store/pull/644#issuecomment-436003138
        return '[Forms] Update Form Value';
    }
}
export class UpdateForm {
    constructor(payload) {
        this.payload = payload;
    }
    static get type() {
        // NOTE: Not necessary to declare the type in this way in your code. See https://github.com/ngxs/store/pull/644#issuecomment-436003138
        return '[Forms] Update Form';
    }
}
export class UpdateFormDirty {
    constructor(payload) {
        this.payload = payload;
    }
    static get type() {
        // NOTE: Not necessary to declare the type in this way in your code. See https://github.com/ngxs/store/pull/644#issuecomment-436003138
        return '[Forms] Update Form Dirty';
    }
}
export class SetFormDirty {
    constructor(payload) {
        this.payload = payload;
    }
    static get type() {
        // NOTE: Not necessary to declare the type in this way in your code. See https://github.com/ngxs/store/pull/644#issuecomment-436003138
        return '[Forms] Set Form Dirty';
    }
}
export class SetFormPristine {
    constructor(payload) {
        this.payload = payload;
    }
    static get type() {
        // NOTE: Not necessary to declare the type in this way in your code. See https://github.com/ngxs/store/pull/644#issuecomment-436003138
        return '[Forms] Set Form Pristine';
    }
}
export class UpdateFormErrors {
    constructor(payload) {
        this.payload = payload;
    }
    static get type() {
        // NOTE: Not necessary to declare the type in this way in your code. See https://github.com/ngxs/store/pull/644#issuecomment-436003138
        return '[Forms] Update Form Errors';
    }
}
export class SetFormDisabled {
    constructor(payload) {
        this.payload = payload;
    }
    static get type() {
        // NOTE: Not necessary to declare the type in this way in your code. See https://github.com/ngxs/store/pull/644#issuecomment-436003138
        return '[Forms] Set Form Disabled';
    }
}
export class SetFormEnabled {
    constructor(payload) {
        this.payload = payload;
    }
    static get type() {
        // NOTE: Not necessary to declare the type in this way in your code. See https://github.com/ngxs/store/pull/644#issuecomment-436003138
        return '[Forms] Set Form Enabled';
    }
}
export class ResetForm {
    constructor(payload) {
        this.payload = payload;
    }
    static get type() {
        // NOTE: Not necessary to declare the type in this way in your code. See https://github.com/ngxs/store/pull/644#issuecomment-436003138
        return '[Forms] Reset Form';
    }
}
